trie/trie.go

```diff
@@ -2,47 +2,47 @@ package trie
 
 import (
 	"bytes"
-	"container/list"
 	"fmt"
 	"sync"
 
 	"github.com/ethereum/go-ethereum/common"
-	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/hashicorp/golang-lru"
 )
 
-func ParanoiaCheck(t1 *Trie, backend Backend) (bool, *Trie) {
-	t2 := New(nil, backend)
+const lruSize = 200
 
+func ParanoiaCheck(t1 *Trie) (bool, *Trie) {
+	t2 := New(common.Hash{}, nil)
 	it := t1.Iterator()
 	for it.Next() {
 		t2.Update(it.Key, it.Value)
 	}
-
-	return bytes.Equal(t2.Hash(), t1.Hash()), t2
+	return t2.Hash() == t1.Hash(), t2
 }
 
 type Trie struct {
-	mu       sync.Mutex
-	root     Node
-	roothash []byte
-	cache    *Cache
-
-	revisions *list.List
+	mu    sync.Mutex
+	root  node
+	db    Backend
+	cache *lru.Cache
+	*hasher
 }
 
-func New(root []byte, backend Backend) *Trie {
+func New(root common.Hash, backend Backend) *Trie {
 	trie := &Trie{}
-	trie.revisions = list.New()
-	trie.roothash = root
 	if backend != nil {
-		trie.cache = NewCache(backend)
+		trie.db = backend
+		trie.cache, _ = lru.New(lruSize)
 	}
-
-	if root != nil {
-		value := common.NewValueFromBytes(trie.cache.Get(root))
-		trie.root = trie.mknode(value)
+	if (root != common.Hash{}) {
+		if backend == nil {
+			panic("cannot use exisiting root without a database")
+		}
+		// if v, _ := trie.db.Get(root[:]); len(v) == 0 {
```

Comment:
I added this for debugging. Please note that the PR is not done yet.If it is uncommented, several tests panic.